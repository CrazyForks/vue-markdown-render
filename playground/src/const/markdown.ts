export // 模拟流式传输
const streamContent = `>>>I'll create a simple Electron + Vue chat application demo. Here's the structure:

[Star on GitHub](https://github.com/Simon-He95/vue-markdown-render)

![Vue Markdown Icon](/vue-markdown-icon.svg "Vue Markdown Icon")
好的，我将为您输出泰勒公式的一般形式及其常见展开式。

---

## 1. 泰勒公式（Taylor's Formula）

### 一般形式（在点 \\(x = a\\) 处展开）：
\\[
f(x) = f(a) + f'(a)(x-a) + \\frac{f''(a)}{2!}(x-a)^2 + \\frac{f'''(a)}{3!}(x-a)^3 + \\cdots + \\frac{f^{(n)}(a)}{n!}(x-a)^n + R_n(x)
\\]

其中：
- \\(f^{(k)}(a)\\) 是 \\(f(x)\\) 在 \\(x=a\\) 处的 \\(k\\) 阶导数
- \\(R_n(x)\\) 是余项，常见形式有拉格朗日余项：
\\[
R_n(x) = \\frac{f^{(n+1)}(\\xi)}{(n+1)!}(x-a)^{n+1}, \\quad \\xi \\text{ 在 } a \\text{ 和 } x \\text{ 之间}
\\]

---

## 2. 麦克劳林公式（Maclaurin's Formula，即 \\(a=0\\) 时的泰勒公式）：
\\[
f(x) = f(0) + f'(0)x + \\frac{f''(0)}{2!}x^2 + \\frac{f'''(0)}{3!}x^3 + \\cdots + \\frac{f^{(n)}(0)}{n!}x^n + R_n(x)
\\]

---

## 3. 常见函数的麦克劳林展开（前几项）

- **指数函数**：
\\[
e^x = 1 + x + \\frac{x^2}{2!} + \\frac{x^3}{3!} + \\cdots + \\frac{x^n}{n!} + \\cdots, \\quad x \\in \\mathbb{R}
\\]

- **正弦函数**：
\\[
\\sin x = x - \\frac{x^3}{3!} + \\frac{x^5}{5!} - \\frac{x^7}{7!} + \\cdots + (-1)^n \\frac{x^{2n+1}}{(2n+1)!} + \\cdots
\\]

- **余弦函数**：
\\[
\\cos x = 1 - \\frac{x^2}{2!} + \\frac{x^4}{4!} - \\frac{x^6}{6!} + \\cdots + (-1)^n \\frac{x^{2n}}{(2n)!} + \\cdots
\\]

- **自然对数**（在 \\(x=0\\) 附近）：
\\[
\\ln(1+x) = x - \\frac{x^2}{2} + \\frac{x^3}{3} - \\frac{x^4}{4} + \\cdots + (-1)^{n-1} \\frac{x^n}{n} + \\cdots, \\quad -1 < x \\le 1
\\]

- **二项式展开**（\\( (1+x)^m \\)，\\(m\\) 为实数）：
\\[
(1+x)^m = 1 + mx + \\frac{m(m-1)}{2!}x^2 + \\frac{m(m-1)(m-2)}{3!}x^3 + \\cdots, \\quad |x| < 1
\\]

---

如果您需要某个特定函数在特定点的泰勒展开，请告诉我，我可以为您详细写出。
*Figure: Vue Markdown Icon (served from /vue-markdown-icon.svg)*

::: warning
这是一个警告块。
:::

::: tip 提示标题
这是带标题的提示。
:::

::: error 错误块
这是一个错误块。
:::

مرحبا بكم في عالم اللغة العربية!
1. First, let's set up the project:

\`\`\`shellscript
# Create Vue project
npm create vue@latest electron-vue-chat

# Navigate to project
cd electron-vue-chat

# Install dependencies
npm install
npm install electron electron-builder vue-router

# Install dev dependencies
npm install -D electron-dev-server concurrently wait-on
\`\`\`

2. Create the main Electron file:

\`\`\`javascript:electron/main.js
const { app, BrowserWindow } = require('electron');
const path = require('path');
const isDev = process.env.NODE_ENV === 'development';

let mainWindow;

function createWindow() {
  mainWindow = new BrowserWindow({
    width: 900,
    height: 680,
    webPreferences: {
      nodeIntegration: true,
      contextIsolation: false
    }
  });

  const url = isDev
    ? 'http://localhost:5173'
    : \`file://\${path.join(__dirname, '../dist/index.html')}\`;

  mainWindow.loadURL(url);

  if (isDev) {
    mainWindow.webContents.openDevTools();
  }

  mainWindow.on('closed', () => {
    mainWindow = null;
  });
}

app.on('ready', createWindow);

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  if (mainWindow === null) {
    createWindow();
  }
});
\`\`\`

3. Update package.json:

\`\`\`diff json:package.json
{
  "name": "vue-renderer-markdown",
  "type": "module",
- "version": "0.0.49",
+ "version": "0.0.54-beta.1",
  "packageManager": "pnpm@10.16.1",
  "description": "A Vue 3 component that renders Markdown string content as HTML, supporting custom components and advanced markdown features.",
  "author": "Simon He",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git + git@github.com:Simon-He95/vue-markdown-render.git"
  },
  "bugs": {
    "url": "https://github.com/Simon-He95/vue-markdown-render/issues"
  },
  "keywords": [
    "vue",
    "vue3",
    "markdown",
    "markdown-to-html",
    "markdown-renderer",
    "vue-markdown",
    "vue-component",
    "html",
    "renderer",
    "custom-component"
  ],
  "exports": {
    ".": {
      "types": "./dist/types/exports.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./index.css": "./dist/index.css",
    "./index.tailwind.css": "./dist/index.tailwind.css",
    "./tailwind": "./dist/tailwind.ts"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/types/exports.d.ts",
  "files": [
    "dist"
  ],
}
\`\`\`

4. Create chat components (diversified languages):

\`\`\`python:src/server/app.py
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Message(BaseModel):
    sender: str
    text: str

@app.get("/health")
def health():
    return {"status": "ok"}

@app.post("/echo")
def echo(msg: Message):
    return {"reply": f"Echo: {msg.text}"}
\`\`\`

\`\`\`tsx:src/components/ChatWindow.tsx
import React, { useEffect, useRef, useState } from 'react';

type Message = { sender: 'me' | 'other'; text: string; time: string };

export default function ChatWindow() {
  const [messages, setMessages] = useState<Message[]>([
    { sender: 'other', text: 'Hi there!', time: '10:00' },
  ]);
  const [input, setInput] = useState('');
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const el = containerRef.current;
    if (!el) return;
    el.scrollTop = el.scrollHeight;
  }, [messages]);

  const send = () => {
    const text = input.trim();
    if (!text) return;
    const now = new Date();
    const time = now.getHours() + ':' + String(now.getMinutes()).padStart(2,'0');
    setMessages((m) => [...m, { sender: 'me', text, time }]);
    setInput('');
    setTimeout(() => {
      setMessages((m) => [...m, { sender: 'other', text: 'Auto-reply', time }]);
    }, 800);
  };

  return (
    <div className="chat-window">
      <div className="messages" ref={containerRef}>
        {messages.map((m, i) => (
          <div key={i} className={'msg ' + m.sender}>
            {m.text}
            <span className="time">{m.time}</span>
          </div>
        ))}
      </div>
      <div className="input">
        <input
          value={input}
          onChange={e => setInput(e.target.value)}
          onKeyDown={e => e.key==='Enter' && send()}
          placeholder="Type..."
        />
        <button onClick={send}>Send</button>
      </div>
    </div>
  );
}
\`\`\`

5. Create a native module example (C++):

\`\`\`cpp:src/native/compute.cpp
#include <bits/stdc++.h>
using namespace std;

int fibonacci(int n){
  if(n<=1) return n;
  int a=0,b=1;
  for(int i=2;i<=n;++i){ int c=a+b; a=b; b=c; }
  return b;
}

int main(){
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout << "fib(10)=" << fibonacci(10) << "\n";
  return 0;
}
\`\`\`

6. Update the main App.vue:

\`\`\`vue:src/App.vue
<template>
  <router-view />
<\/template>

<style>
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  font-family: Arial, sans-serif;
}
</style>
\`\`\`

7. Set up the router:

\`\`\`javascript:src/router/index.js
import { createRouter, createWebHistory } from 'vue-router';
import ChatView from '../views/ChatView.vue';

const routes = [
  {
    path: '/',
    name: 'chat',
    component: ChatView
  }
];

const router = createRouter({
  history: createWebHistory(),
  routes
});

export default router;
\`\`\`

8. Update main.js:

\`\`\`javascript:src/main.js
import { createApp } from 'vue';
import App from './App.vue';
import router from './router';

createApp(App).use(router).mount('#app');
\`\`\`

To run the application:

\`\`\`bash
npm run electron:dev
\`\`\`

This creates a simple chat application with a contacts sidebar and a chat window. Users can select contacts and send messages, with auto-replies simulated.


9. 付费转化率计算验证:
1. **公式**
   \[
   \text{付费转化率} = \\left( \\frac{\text{付费用户数}}{\text{月活用户数}} \\right) \times 100\%
   \]

2. **代入数据**
   \[
   \\frac{363}{15,\!135} \times 100\% = 2.398\%
   \]

3. **计算工具验证**
   通过数学计算工具确认结果：
   \`363 ÷ 15,135 × 100 = 2.39841427...\`

4. **差异说明**
   $$E=mc^2$$

10. Mermaid graphic:

\`\`\`mermaid
graph TD
    Kira_Yamato[基拉·大和]
    Lacus_Clyne[拉克丝·克莱因]
    Athrun_Zala[阿斯兰·萨拉]
    Cagalli_Yula_Athha[卡嘉莉·尤拉·阿斯哈]
    Shinn_Asuka[真·飞鸟]
    Lunamaria_Hawke[露娜玛丽亚·霍克]
    COMPASS[世界和平监视组织COMPASS]
    Foundation[芬德申王国]
    Orphee_Lam_Tao[奥尔菲·拉姆·陶]
    %% 节点定义结束，开始定义边
    Kira_Yamato ---|恋人| Lacus_Clyne
    Kira_Yamato ---|挚友| Athrun_Zala
    Kira_Yamato -->|隶属| COMPASS
    Kira_Yamato -->|前辈| Shinn_Asuka
    Lacus_Clyne -->|初代总裁| COMPASS
    Athrun_Zala ---|恋人| Cagalli_Yula_Athha
    Athrun_Zala -.->|协力| COMPASS
    Shinn_Asuka ---|恋人| Lunamaria_Hawke
    Shinn_Asuka -->|隶属| COMPASS
    Lunamaria_Hawke -->|隶属| COMPASS
    COMPASS -->|对立| Foundation
    Orphee_Lam_Tao -->|隶属| Foundation
    Orphee_Lam_Tao -.->|追求| Lacus_Clyne
\`\`\`

`
