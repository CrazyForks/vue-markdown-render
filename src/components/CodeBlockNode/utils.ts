import { defineAsyncComponent } from 'vue'

let IconifyComponent = null
export function getIconify() {
  if (IconifyComponent)
    return IconifyComponent
  IconifyComponent = defineAsyncComponent(async () => {
    try {
      const mod = await import('@iconify/vue')
      return (mod as any).Icon
    }
    catch {
      throw new Error('Optional dependency "@iconify/vue" is not installed. Please install it to enable language icons in code blocks.')
    }
  })
  return IconifyComponent
}

let mod = null
export async function getUseMonaco() {
  if (mod)
    return mod
  try {
    mod = await import('vue-use-monaco')
  }
  catch {
    throw new Error('Optional dependency "vue-use-monaco" is not installed. Please install it to enable code editor features.')
  }
  return mod
}
