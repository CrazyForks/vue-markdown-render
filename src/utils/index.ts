export * from './code.detect'
export * from './code.theme'
export * from './theme'
export * from './useCodeEditor'
export * from './cn'
export * from './markdown'
