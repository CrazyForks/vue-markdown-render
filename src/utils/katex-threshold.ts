/**
 * Recommend when to use Worker based on formula count and complexity.
 *
 * N_threshold ≈ floor(B / (R × (1 - H)))
 * - B: main-thread budget (ms), e.g. 50ms or a stricter 16.7ms
 * - R: avg render time per unique formula (ms)
 * - H: cache hit rate (0~1)
 */

export interface ThresholdInput {
  R: number // avg render time per unique formula, ms
  H?: number // cache hit rate (0~1), default 0 for first paint
  B?: number // main-thread budget ms, default 50
}

export function recommendWorkerThreshold({ R, H = 0, B = 50 }: ThresholdInput) {
  const denom = R * (1 - H) || 1e-6
  const N = Math.max(1, Math.floor(B / denom))
  return N
}

// A very lightweight classifier by formula length/characters to pick R ballpark
export function estimateRByFormula(sample: string): 'simple' | 'medium' | 'complex' {
  const len = sample.length
  const slashes = (sample.match(/\\/g) || []).length
  // heuristics: more backslashes and longer length indicate complexity
  const score = len + slashes * 10
  if (score < 10)
    return 'simple'
  if (score < 40)
    return 'medium'
  return 'complex'
}

export function defaultRByClass(cls: 'simple' | 'medium' | 'complex'): number {
  // conservative defaults (ms) for mid-tier devices
  switch (cls) {
    case 'simple': return 3
    case 'medium': return 10
    case 'complex': return 30
  }
}

export function recommendNForSamples(formulas: string[], opts?: { H?: number, B?: number }) {
  // classify each and take the worst (max R) since bursts often contain mixed complexity
  let maxR = 0
  for (const f of formulas) {
    const cls = estimateRByFormula(f)
    const R = defaultRByClass(cls)
    if (R > maxR)
      maxR = R
  }
  return recommendWorkerThreshold({ R: maxR, H: opts?.H ?? 0, B: opts?.B ?? 50 })
}
