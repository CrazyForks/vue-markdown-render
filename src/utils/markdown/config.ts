/**
 * MathOptions control how the math plugin normalizes content before
 * handing it to KaTeX (or other math renderers).
 *
 * - commands: list of command words that should be auto-prefixed with a
 *   backslash if not already escaped (e.g. 'infty' -> '\\infty'). Use a
 *   conservative list to avoid false positives in prose.
 * - escapeExclamation: whether to escape standalone '!' to '\\!' (default true).
 */
export interface MathOptions {
  /** List of command words to auto-escape. */
  commands?: readonly string[]
  /** Whether to escape standalone '!' (default: true). */
  escapeExclamation?: boolean
}

let defaultMathOptions: MathOptions | undefined

export function setDefaultMathOptions(opts: MathOptions | undefined) {
  defaultMathOptions = opts
}

export function getDefaultMathOptions(): MathOptions | undefined {
  return defaultMathOptions
}
