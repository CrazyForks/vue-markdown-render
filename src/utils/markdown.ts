import katex from 'katex'
import MarkdownIt from 'markdown-it'
import { full as markdownItEmoji } from 'markdown-it-emoji'
import markdownItFootnote from 'markdown-it-footnote'
import markdownItIns from 'markdown-it-ins'
import markdownItMark from 'markdown-it-mark'
import markdownItSub from 'markdown-it-sub'
import markdownItSup from 'markdown-it-sup'
import * as markdownItCheckbox from 'markdown-it-task-checkbox'
import { useSafeI18n } from '../composables/useSafeI18n'
import { renderKaTeXInWorker, setKaTeXCache } from '../workers/katexWorkerClient'

import {
  parseInlineTokens,
  parseMarkdownToStructure,
  processTokens,
} from './markdown-parser'
import { getMarkdown as factory } from './markdown/getMarkdown'
import 'katex/dist/katex.min.css'

// Re-export the node types for backward compatibility
export * from '../types'
export { parseInlineTokens, parseMarkdownToStructure, processTokens }

export interface GetMarkdownOptions {
  plugin?: Array<any>
  apply?: Array<(md: MarkdownIt) => void>
}

export function getMarkdown(msgId: string = `editor-${Date.now()}`, options: GetMarkdownOptions = {}) {
  // keep legacy behaviour but delegate to new factory and reapply project-specific rules
  const md = factory({})

  // apply user supplied plugins (md.use)
  if (Array.isArray(options.plugin)) {
    for (const p of options.plugin) {
      // allow both [plugin, opts] tuple or plugin function
      if (Array.isArray(p))
        md.use(p[0], p[1])
      else
        md.use(p)
    }
  }

  // apply user supplied apply functions to mutate the md instance (e.g. md.block.ruler.before(...))
  if (Array.isArray(options.apply)) {
    for (const fn of options.apply) {
      try {
        fn(md)
      }
      catch (e) {
        // swallow errors to preserve legacy behaviour; developers can see stack in console

        console.error('[getMarkdown] apply function threw an error', e)
      }
    }
  }

  // Re-apply a few project specific plugins that were previously always enabled
  md.use(markdownItSub)
  md.use(markdownItSup)
  md.use(markdownItMark)
  md.use(markdownItEmoji)
  const markdownItCheckboxPlugin
    = (markdownItCheckbox as any).default ?? markdownItCheckbox
  md.use(markdownItCheckboxPlugin)
  md.use(markdownItIns)
  md.use(markdownItFootnote)

  // Annotate fence tokens with unclosed meta using a lightweight line check
  md.core.ruler.after('block', 'mark_fence_closed', (state: any) => {
    const src: string = state.src as string
    const lines = src.split(/\r?\n/)
    for (const token of state.tokens) {
      if (token.type !== 'fence' || !token.map || !token.markup)
        continue
      const openLine: number = token.map[0]
      const endLine: number = token.map[1]
      const markup: string = token.markup
      const marker = markup[0]
      const minLen = markup.length
      // The closing line, if exists, should be the last line consumed by the block
      const lineIdx = Math.max(0, endLine - 1)
      const line = lines[lineIdx] ?? ''
      let i = 0
      while (i < line.length && (line[i] === ' ' || line[i] === '\t')) i++
      let count = 0
      while (i + count < line.length && line[i + count] === marker) count++
      let j = i + count
      while (j < line.length && (line[j] === ' ' || line[j] === '\t')) j++
      const closed = endLine > openLine + 1 && count >= minLen && j === line.length
      token.meta = token.meta || {}
      token.meta.unclosed = !closed
      // also set a explicit `closed` boolean for compatibility with plugins/tests
      token.meta.closed = !!closed
    }
  })

  // wave rule (legacy)
  const waveRule = (state: any, silent: boolean) => {
    const start = state.pos
    if (state.src[start] !== '~')
      return false
    const prevChar = state.src[start - 1]
    const nextChar = state.src[start + 1]
    if (/\d/.test(prevChar) && /\d/.test(nextChar)) {
      if (!silent) {
        const token = state.push('text', '', 0)
        token.content = '~'
      }
      state.pos += 1
      return true
    }
    return false
  }

  md.inline.ruler.before('sub', 'wave', waveRule)

  // custom fence that uses msgId for unique ids
  md.renderer.rules.fence = (tokens: any, idx: number) => {
    const { t } = useSafeI18n()
    const token = tokens[idx]
    const info = token.info ? token.info.trim() : ''
    const str = token.content
    const encodedCode = btoa(unescape(encodeURIComponent(str)))
    const language = info || 'text'
    const uniqueId = `editor-${msgId}-${idx}-${language}`

    return `<div class="code-block" data-code="${encodedCode}" data-lang="${language}" id="${uniqueId}">
      <div class="code-header">
        <span class="code-lang">${language.toUpperCase()}</span>
        <button class="copy-button" data-code="${encodedCode}">${t(
          'common.copy',
        )}</button>
      </div>
      <div class="code-editor"></div>
    </div>`
  }

  // reference rule (legacy)
  const referenceInline = (state: any, silent: boolean) => {
    if (state.src[state.pos] !== '[')
      return false
    const match = /^\[(\d+)\]/.exec(state.src.slice(state.pos))
    if (!match)
      return false
    if (!silent) {
      const id = match[1]
      const token = state.push('reference', 'span', 0)
      token.content = id
      token.markup = match[0]
    }
    state.pos += match[0].length
    return true
  }

  md.inline.ruler.before('escape', 'reference', referenceInline)
  md.renderer.rules.reference = (tokens: any, idx: number) => {
    const id = tokens[idx].content
    return `<span class="reference-link" data-reference-id="${id}" role="button" tabindex="0" title="Click to view reference">${id}</span>`
  }

  return md
}

export function getCommonMarkdown() {
  const md = new MarkdownIt({
    html: true,
    linkify: true,
    typographer: true,
    breaks: false,
  })
  return md
}

export function renderMarkdown(md: MarkdownIt, content: string) {
  const html = md.render(content)

  const newHTML = html.replace(/\([^)]*\)/g, (match) => {
    const latex = match.slice(1, -1) // remove surrounding parentheses
    try {
      const data = katex.renderToString(latex, {
        throwOnError: true,
        displayMode: false,
      })
      return data
    }
    catch {
      return match
    }
  })
  return newHTML
}

// Async version that uses the KaTeX worker (if available) to render inline parentheses content
// to avoid blocking the main thread. Falls back to synchronous renderToString on worker failure.
export async function renderMarkdownAsync(md: MarkdownIt, content: string) {
  const html = md.render(content)

  const matches = [...html.matchAll(/\([^)]*\)/g)]
  if (matches.length === 0)
    return html

  // Kick off all renders in parallel (worker preferred), then apply replacements
  const renderPromises = matches.map((m) => {
    const latex = (m[0] || '').slice(1, -1)
    return (async () => {
      try {
        return await renderKaTeXInWorker(latex, false, 800)
      }
      catch {
        try {
          const data = katex.renderToString(latex, {
            throwOnError: true,
            displayMode: false,
          })
          try {
            setKaTeXCache(latex, false, data)
          }
          catch {
            // ignore cache set errors
          }
          return data
        }
        catch {
          return null
        }
      }
    })()
  })

  const results = await Promise.all(renderPromises)

  // Apply replacements in reverse order so indices remain valid
  let result = html
  for (let i = matches.length - 1; i >= 0; i--) {
    const m = matches[i]
    const start = m.index ?? 0
    const end = start + m[0].length
    const rendered = results[i]
    if (rendered)
      result = result.slice(0, start) + rendered + result.slice(end)
  }

  return result
}
