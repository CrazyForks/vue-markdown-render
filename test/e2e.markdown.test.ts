import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import { getMarkdown, parseMarkdownToStructure, renderMarkdown } from '../src/utils/markdown'

const md = getMarkdown('e2e')

const FIXTURES_DIR = resolve(__dirname, 'fixtures')

describe('e2e markdown parsing (fixtures)', () => {
  const files = [
    'headings.md',
    'code-diff.md',
    'table.md',
    'admonition.md',
    'math.md',
    'footnotes.md',
    'image-link.md',
    'checkbox.md',
    'mermaid.md',
    // edge-case fixtures
    'unclosed-fence.md',
    'trailing-backticks.md',
    'unmatched-brackets.md',
    'escaped-brackets.md',
    'nested-lists-edge.md',
    'fence-with-meta.md',
  ]

  for (const f of files) {
    it(`parses fixture ${f}`, () => {
      const src = readFileSync(resolve(FIXTURES_DIR, f), 'utf8')
      const nodes = parseMarkdownToStructure(src, md)
      expect(nodes).toBeInstanceOf(Array)
      // Ensure we have at least one top-level node
      expect(nodes.length).toBeGreaterThan(0)

      // Create a stable, minimal serialization for snapshot
      // Helper to find first descendant text for a node
      const findFirstText = (node: any): string | undefined => {
        if (!node)
          return undefined
        if (node.type === 'text' && typeof node.content === 'string')
          return node.content.slice(0, 80)
        if (Array.isArray(node.children)) {
          for (const c of node.children) {
            const t = findFirstText(c)
            if (t)
              return t
          }
        }
        return undefined
      }

      const pickMeta = (meta: any) => {
        if (!meta || typeof meta !== 'object')
          return undefined
        const keys = ['diff', 'language', 'closed', 'unclosed']
        const out: any = {}
        for (const k of keys) {
          if (k in meta)
            out[k] = meta[k]
        }
        return Object.keys(out).length ? out : undefined
      }

      const minimal = nodes.map((n: any) => {
        const out: any = { type: n.type }
        out.firstText = findFirstText(n)
        if (n.meta)
          out.meta = pickMeta(n.meta)

        if (n.type === 'heading') {
          out.level = (n as any).level
          out.children = (n as any).children?.map((c: any) => c.type)
        }
        else if (n.type === 'list') {
          out.ordered = (n as any).ordered
          out.items = (n as any).items?.length ?? 0
        }
        else if (n.type === 'code_block') {
          out.language = (n as any).language
          out.diff = !!(n as any).diff
          // include a short preview of code
          out.preview = (n as any).code ? (n as any).code.slice(0, 120) : undefined
        }
        else if (n.type === 'table') {
          out.rows = (n as any).rows?.length ?? 0
        }
        else if (n.type === 'footnote') {
          out.items = (n as any).items?.length ?? 0
        }
        else {
          // generic: record if has children
          out.children = (n as any).children ? (n as any).children.length : undefined
        }
        return out
      })

      expect(minimal).toMatchSnapshot(f)
    })
  }

  it('renders markdown to HTML containing expected fragments', () => {
    const small = 'This has inline math (\\alpha) and code:\n\n```ts\nlet x = 1\n```'
    const html = renderMarkdown(md, small)
    expect(html).toContain('code-block')
    // inline math is rendered via KaTeX into a <math> or HTML; assert a math-like fragment
    expect(html).toMatch(/<math|katex|\^|\\alpha/)
  })
})
