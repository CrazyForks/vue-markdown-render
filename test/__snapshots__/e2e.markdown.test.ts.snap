// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e markdown parsing (fixtures) > parses fixture admonition.md > admonition.md 1`] = `
[
  {
    "children": 1,
    "firstText": "This is a warning admonition with ",
    "type": "admonition",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture checkbox.md > checkbox.md 1`] = `
[
  {
    "firstText": undefined,
    "items": 2,
    "ordered": false,
    "type": "list",
  },
  {
    "children": 1,
    "firstText": "Regular list item",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture code-diff.md > code-diff.md 1`] = `
[
  {
    "diff": true,
    "firstText": undefined,
    "language": "",
    "preview": "@@
  new line
",
    "type": "code_block",
  },
  {
    "diff": false,
    "firstText": undefined,
    "language": "js",
    "preview": "console.log('hello world')
",
    "type": "code_block",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture escaped-brackets.md > escaped-brackets.md 1`] = `
[
  {
    "children": 1,
    "firstText": "A line with an escaped bracket: [not a link]",
    "type": "paragraph",
  },
  {
    "children": 1,
    "firstText": "And a literal backslash \\ and more text.",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture fence-with-meta.md > fence-with-meta.md 1`] = `
[
  {
    "diff": true,
    "firstText": undefined,
    "language": "js",
    "preview": "@@
  new
",
    "type": "code_block",
  },
  {
    "diff": false,
    "firstText": undefined,
    "language": "python",
    "preview": "# regular code block
print('ok')
",
    "type": "code_block",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture footnotes.md > footnotes.md 1`] = `
[
  {
    "children": 3,
    "firstText": "This is a sentence with a footnote",
    "type": "paragraph",
  },
  {
    "firstText": "This is the footnote text.",
    "items": 0,
    "type": "footnote",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture headings.md > headings.md 1`] = `
[
  {
    "children": [
      "text",
    ],
    "firstText": "Heading 1",
    "level": 1,
    "type": "heading",
  },
  {
    "children": 1,
    "firstText": "A short paragraph under heading 1.",
    "type": "paragraph",
  },
  {
    "children": [
      "text",
    ],
    "firstText": "Heading 2",
    "level": 2,
    "type": "heading",
  },
  {
    "children": 5,
    "firstText": "Another paragraph with ",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture image-link.md > image-link.md 1`] = `
[
  {
    "children": 2,
    "firstText": "Here is an image: ",
    "type": "paragraph",
  },
  {
    "children": 2,
    "firstText": "And a link: ",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture math.md > math.md 1`] = `
[
  {
    "children": 5,
    "firstText": "Inline math ",
    "type": "paragraph",
  },
  {
    "children": undefined,
    "firstText": undefined,
    "type": "math_block",
  },
  {
    "children": 2,
    "firstText": "A line with parentheses that should not break: ",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture mermaid.md > mermaid.md 1`] = `
[
  {
    "diff": false,
    "firstText": undefined,
    "language": "mermaid",
    "preview": "graph TD
  A-->B
  B-->C
",
    "type": "code_block",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture nested-lists-edge.md > nested-lists-edge.md 1`] = `
[
  {
    "firstText": undefined,
    "items": 2,
    "ordered": true,
    "type": "list",
  },
  {
    "firstText": undefined,
    "items": 1,
    "ordered": false,
    "type": "list",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture table.md > table.md 1`] = `
[
  {
    "firstText": undefined,
    "rows": 2,
    "type": "table",
  },
  {
    "children": 1,
    "firstText": "A paragraph after the table.",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture trailing-backticks.md > trailing-backticks.md 1`] = `
[
  {
    "diff": false,
    "firstText": undefined,
    "language": "js",
    "preview": "const a = 1
",
    "type": "code_block",
  },
  {
    "children": 1,
    "firstText": undefined,
    "type": "paragraph",
  },
  {
    "children": 1,
    "firstText": "Some text after an extra backtick line.",
    "type": "paragraph",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture unclosed-fence.md > unclosed-fence.md 1`] = `
[
  {
    "diff": false,
    "firstText": undefined,
    "language": "js",
    "preview": "console.log('unclosed fence')

# Following heading that might be part of the fence
",
    "type": "code_block",
  },
]
`;

exports[`e2e markdown parsing (fixtures) > parses fixture unmatched-brackets.md > unmatched-brackets.md 1`] = `
[
  {
    "children": 1,
    "firstText": "This line has an opening bracket [ without a closing.",
    "type": "paragraph",
  },
  {
    "children": 2,
    "firstText": "Another line with a link-like ",
    "type": "paragraph",
  },
  {
    "children": 1,
    "firstText": "And a normal line.",
    "type": "paragraph",
  },
]
`;
